tempodb/encoding/vparquet2/nested_set_model.go

```diff
@@ -0,0 +1,154 @@
+package vparquet2
+
+import (
+	v1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+	"github.com/grafana/tempo/pkg/util"
+)
+
+// spanNode is a wrapper around a span that is used to build and travers spans as a tree.
+type spanNode struct {
+	parent    *spanNode
+	span      *Span
+	children  []*spanNode
+	nextChild int
+}
+
+// assignNestedSetModelBounds calculates and assigns the values Span.NestedSetLeft, Span.NestedSetRight,
+// and Span.ParentID for all spans in a trace. The assignment is skipped when all spans have non-zero
+// left and right bounds. If forceAssignment is true, the assignment is never skipped.
+func assignNestedSetModelBounds(trace *Trace, forceAssignment bool) {
+	// count spans in order be able to pre-allocate tree nodes
+	var spanCount int
+	for _, rs := range trace.ResourceSpans {
+		for _, ss := range rs.ScopeSpans {
+			spanCount += len(ss.Spans)
+		}
+	}
+
+	// find root spans and map span IDs to tree nodes
+	var (
+		undoAssignment   bool
+		assignmentNeeded bool
+		allNodes         = make([]spanNode, 0, spanCount)
+		nodesByID        = make(map[uint64][]*spanNode, spanCount)
+		rootNodes        []*spanNode
+	)
+
+	for _, rs := range trace.ResourceSpans {
+		for _, ss := range rs.ScopeSpans {
+			for i, s := range ss.Spans {
+				if s.NestedSetLeft == 0 || s.NestedSetRight == 0 {
+					assignmentNeeded = true
+				}
+
+				allNodes = append(allNodes, spanNode{span: &ss.Spans[i]})
+				node := &allNodes[len(allNodes)-1]
+
+				if s.IsRoot() {
+					rootNodes = append(rootNodes, node)
+				}
+
+				id := util.SpanIDToUint64(s.SpanID)
+				if nodes, ok := nodesByID[id]; ok {
+					// zipkin traces may contain client/server spans with the same IDs
+					nodesByID[id] = append(nodes, node)
+					if len(nodes) > 2 {
+						undoAssignment = true
+					}
+				} else {
+					nodesByID[id] = []*spanNode{node}
+				}
+			}
+		}
+	}
+
+	// check preconditions before assignment
+	if undoAssignment {
```

Comment:
This is for the case where more than two spans have the same ID and it is no longer possible to make a reasonable assignment. In case there are nested set bounds assigned, it is necessary to undo them.